DllImportGenerator/designs/StructMarshalling.md

```diff
@@ -0,0 +1,225 @@
+  
+## Possible Solution A - Opt-In Structure Interop
+
+We've been working around another problem for a while in the runtime-integrated interop design: The user can use any type that is non-auto layout and has fields that can be marshaled in interop. This has lead to various issues where types that were not intended for interop usage became usable and then we couldn't update their behavior to be special-cased since users may have been relying on the generic behavior (`Span<T>`, `Vector<T>` come to mind).
+
+I propose an opt-in design where the owner of a struct has to explicitly opt-in to usage for interop. This enables our team to add special support as desired for various types such as `Span<T>` while also avoiding the private reflection and limited type information issues mentioned above.
+
+This design would use these attributes:
+
+```csharp
+
+[AttributeUsage(AttributeTargets.Struct | AttributeTargets.Class)]
+public class GeneratedMarshallingAttribute : Attribute {}
+
+[AttributeUsage(AttributeTargets.Struct | AttributeTargets.Class)]
+public class BlittableTypeAttribute : Attribute {}
+
+[AttributeUsage(AttributeTargets.Struct | AttributeTargets.Class)]
+public class NativeMarshallingAttribute : Attribute
+{
+     public NativeMarshallingAttribute(Type nativeType) {}
+}
+
+[AttributeUsage(AttributeTargets.Parameter | AttributeTargets.ReturnValue | AttributeTargets.Field)]
+public class MarshalUsingAttribute : Attribute
+{
+     public MarshalUsingAttribute(Type nativeType) {}
+}
+
+```
+
+The `NativeMarshallingAttribute` and `MarshalUsingAttribute` attributes would require that the provided native type `TNative` is a blittable `struct` and has three methods with the following names and shapes (with the managed type named TManaged):
+
+```csharp
+struct TNative
+{
+     public TNative(TManaged managed)
+     {}
+
+     public TManaged ToManaged() {}
+
+     public void FreeNative() {}
+}
+```
+
+> :question: Does this API surface and shape work for all marshalling scenarios we plan on supporting? It may have issues with the current "layout class" by-value `[Out]` parameter marshalling where the runtime updates a `class` typed object in place. We already recommend against using classes for interop for performance reasons and a struct value passed via `ref` or `out` with the same members would cover this scenario.
+
+If the native type `TNative` also has a public `Value` property, then the value of the `Value` property will be passed to native code instead of the `TNative` value itself. As a result, the type `TNative` will be allowed to be non-blittable and the type of the `Value` property will be required to be blittable.
+
+> :question: Should we support opting into marshalling via pinning by detecting a `GetPinnableReference` method with a blittable ref return type as an optional addition to the support of `NativeMarshallingAttribute` and `MarshalUsingAttribute`?
```

Comment:
> GetPinnableReference

I think so.

Another similar case can be stackalloc buffers. They have to be provided from the outside (unless we invent some new intrinsic way  to allocate buffers on stack).

Reply:
We would only use stackalloc buffers for types that we would most likely special case in the generator: arrays, strings, and spans. I don't think it would make sense for any user types to opt-in to being allocated via stackalloc instead of just via stack variables.

Reply:
The specialization makes sense for the most performance sensitive types. Once there are data transformations involved, it is slower by design. I do not think it is necessary to special case it for performance. Ideally, it should use regular mechanisms.

For example, it would be nice to be able to marshal strings via a custom encoding (e.g. UTF-32), with the same performance as regular strings.

Reply:
I'm trying to think of a good way to have a native type opt-in to allocating memory via stackalloc but I'm struggling.

We can't have an `Allocate` method on the native type because that breaks the stackalloc scope.

We could put it on the `NativeMarshalling` or `MarshalUsing` attributes, but then for the UTF-32 string case, it would have to be specified at each usage site.

Any other ideas that I'm thinking of become increasingly obtuse.


Do you have any ideas?

Reply:
I was thinking about something like: If the transport type constructor has a second `Span<char> stackBuffer argument`, the source generator is required to provide stackalloced buffer to it:

```
public struct UTF32StringMarshaler
{
    public UTF32StringMarshaler(string s, Span<char> stackBuffer);

    public const StackBufferSize = 0x100;
}

...
static partial void MyPInvoke([MarshalUsing(UTF32StringMarshaler)] string s);
```

generates:

```
static partial void MyPInvoke(string s)
{
    UTF32Marshaler arg0 = new UTF32StringMarshaler(s, stackalloc char[UTF32Marshaler.StackBufferSize]);
...
}
```